### ISSUE TITLE
Implement BMPT `ParallelStateRoot`

### ISSUE THREAD

**Issue by frisitano** [scroll-tech/reth#72]
### Describe the feature

Currently, we only have a `StateRoot` implementation for the BMPT (#36). However, in live sync mode, the engine uses a `ParallelStateRoot` object, which computes all storage tries in parallel. We should create a `ParallelStateRoot` type for the BMPT, introduce a `ParallelStateRootProvider`, and include it in the `StateCommitment` type.

### Implementation
First, we should review usage of the `ParallelStateRoot` object within the reth codebase and implement a `DatabaseParallelStateRoot` trait in https://github.com/scroll-tech/reth/blob/scroll/crates/trie/db/src/state.rs. We should then introduce a `ParallelStateRoot` type for scroll in https://github.com/scroll-tech/reth/tree/scroll/crates/scroll/state-commitment/src/root. We should implement `DatabaseParallelStateRoot` on both the scroll and native `ParallelStateRoot` types. We should add `ParallelStateRoot` GAT which is bound by `DatabaseParallelStateRoot` in `StateCommitment`. We should update the concrete implementations of `StateCommitment` for both BMPT and MPT. We should introduce a `ParallelStateRootProvider` in https://github.com/scroll-tech/reth/tree/scroll/crates/storage/provider. Alternatively, it may be sufficient to add additional methods to `StateRootProviderExt` if we only use the `ParallelStateRoot` when calculating the state root for the most recent state (this should be investigated).
https://github.com/scroll-tech/reth/blob/b456d1f8ac1df681b9903c194c4356b63f316cdb/crates/storage/storage-api/src/trie.rs#L46-L78
We should replace direct instantiation and invocation of `ParallelStateRoot` in the codebase with invocations of the provider methods.

**Comment by frisitano** [comment]
@greged93 Can you explore this issue by exploring the open questions raised in the description to see what direction we should take please?

**Comment by greged93** [comment]
I'm running into some cyclic dep issues which might limit the above design. Currently, I have added a `ParallelDatabaseStateRoot` trait, which is implemented on `reth-trie-parallel::ParallelStateRoot` and `reth-scroll-state-commitment::ParallelStateRoot`. Because the methods `incremental_root` and `incremental_root_with_updates` require [various trait bounds](https://github.com/scroll-tech/reth/blob/1996f4e758b09ce95c795b610d06f22e59670054/crates/trie/parallel/src/root.rs#L60-L67), these bounds need to be reflected on the `ParallelStateRoot` GAT of `StateCommitment` (same requirement as described [here](https://github.com/paradigmxyz/reth/pull/11842#discussion_r1810242690)). The traits are defined in `reth-provider`, which means `reth-node-types` ends up importing `reth-provider`, which creates the cyclic dep: `reth-node-types` depends on `reth-provider` depends on `reth-node-types`. 

I think being able to go around this would require a rework on the `ConsistentDbView`, replacing the `Factory` generic into a `Tx`, but might be missing something.

**Comment by frisitano** [comment]
Can you provide references to the code where you have implemented this please so I can have some additional context.

**Comment by greged93** [comment]
[here](https://github.com/scroll-tech/reth/tree/feat/parallel-state-root) is the branch for it.
- [`ParallelStateCommit`](https://github.com/scroll-tech/reth/blob/feat/parallel-state-root/crates/trie/parallel/src/commitment.rs#L5-L18): similar to `StateCommitment`, I had it extracted here to avoid circular dependencies.
- [`ParallelDatabaseStateRoot`](https://github.com/scroll-tech/reth/blob/feat/parallel-state-root/crates/trie/parallel/src/db.rs#L11-L21): as described in the issue.
- [`ParallelDatabaseStateRoot`](https://github.com/scroll-tech/reth/blob/feat/parallel-state-root/crates/trie/parallel/src/root.rs#L230) implementation for MPT `ParallelStateRoot`.
- [`ParallelDatabaseStateRoot`](https://github.com/scroll-tech/reth/blob/feat/parallel-state-root/crates/scroll/state-commitment/src/root/parallel.rs#L226) implementation for BMPT `ParallelStateRoot`.

**Comment by frisitano** [comment]
Right, I see the issue here. I think given that we will be deprecating the tree I would propose that we implement `ParallelStateRoot` for BMPT and then feature gate it's instantiation for now. This avoids having to introduce the `ParallelStateCommite` abstraction and the issues associated with it.

**Comment by greged93** [comment]
adding this suggestion from @frisitano here in case we want to explore an abstraction in the future:
- Introduce a `ConsistentDBViewProviderFactory` which we implement on `ConsistentDBView`.
```rust
pub trait ConsistentDBViewProviderFactory {
    type DB: Database;

    type Provider: DBProvider<Tx = <Self::DB as Database>::TX>;

    fn provider_ro(&self) -> ProviderResult<Self::Provider>;
}
```
- Use this trait as a bound in the various traits implementations:
```rust
pub trait ParallelDatabaseStateRoot<F: ConsistentDBViewProviderFactory> {
    /// Returns a parallel state root instance from the [`ConsistentDbView`].
    fn from_consistent_db_view(view: F) -> Result<Self, ProviderError>;
    /// Calculate the state root in parallel.
    fn incremental_root(self, input: TrieInput) -> Result<B256, ParallelStateRootError>;
    /// Calculate the state root in parallel and returns the trie updates.
    fn incremental_root_with_updates(
        self,
        input: TrieInput,
    ) -> Result<(B256, TrieUpdates), ParallelStateRootError>;
}
```

**Comment by frisitano** [comment]
closed by #100
